include/rocksdb/enum_reflection.h

```diff
@@ -0,0 +1,265 @@
+// created by leipeng at 2019-12-25
+// clang-format off
+#pragma once
+#include "rocksdb/preproc.h"
+#include "rocksdb/slice.h"
+#include <type_traits>
+
+namespace ROCKSDB_NAMESPACE {
+  Slice var_symbol(const char* s);
+
+template<class Enum>
+class EnumValueInit {
+    Enum val;
+public:
+    operator Enum() const { return val; }
+
+    /// set val
+    EnumValueInit& operator-(Enum v) { val = v; return *this; }
+
+    /// absorb the IntRep param
+    template<class IntRep>
+    EnumValueInit& operator=(IntRep) { return *this; }
+};
+
+template<class Enum>
+Slice enum_name(Enum v, const char* unkown = "") {
+  auto names  = enum_all_names ((Enum*)0);
+  auto values = enum_all_values((Enum*)0);
+  for (size_t i = 0; i < names.second; ++i) {
+    if (v == values[i])
+      return names.first[i];
+  }
+  return unkown;
+}
+
+template<class Enum>
+std::string enum_stdstr(Enum v) {
+  auto names  = enum_all_names ((Enum*)0);
+  auto values = enum_all_values((Enum*)0);
+  for (size_t i = 0; i < names.second; ++i) {
+    if (v == values[i])
+      return names.first[i].ToString();
+  }
+  return "unkown:" + (sizeof(Enum) <= sizeof(int)
+                          ? std::to_string((int)v)
+                          : std::to_string((long)v));
+}
+
+template<class Enum>
+const char* enum_cstr(Enum v, const char* unkown = "") {
+  auto names  = enum_all_names ((Enum*)0);
+  auto values = enum_all_values((Enum*)0);
+  for (size_t i = 0; i < names.second; ++i) {
+    if (v == values[i])
+      return names.first[i].data();
+  }
+  return unkown;
+}
+
+template<class Enum>
+bool enum_value(const ROCKSDB_NAMESPACE::Slice& name, Enum* result) {
```

Comment:
Why "Slice" and not std::string" ?

Reply:
`std::string` and `const char*` can be implicitly converted to `Slice`, and `Slice` is a fundamental type in rocksdb.